### PR TITLE
Work towards support for long paths on Windows

### PR DESCRIPTION
**Pull request by jimhester**

### CODE REVIEW

**Comment by jimhester** [comment]
So this fixes the first issue with `_wfopen`, but we then run into a separate issue when setting the mtime for the files on https://github.com/r-lib/zip/blob/master/src/zip.c#L256

**Comment by jimhester** [comment]
``` r
d <- tempfile()
dir.create(d)
out_file <- file.path(d, "master.zip")

download.file("https://github.com/r-lib/devtools/archive/master.zip", out_file)

out_dir <- file.path(tempdir(), paste0(collapse = "", rep("x", 100)))
dir.create(out_dir)
nchar(out_dir)
#> [1] 148
zip::unzip(out_file, exdir = out_dir)
#> Error in zip::unzip(out_file, exdir = out_dir): zip error: `Failed to set mtime on `devtools-master/tests/testthat/shallowRepo/objects/pack/pack-c4e0f1d1d68408f260cbbf0a533ad5f6bfd5524e.pack` while extracting `C:\Users\Administrator\AppData\Local\Temp\RtmpcPZdkd\file152457c057ae\master.zip`` in file `zip.c:260`
```

<sup>Created on 2020-08-31 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This is the second error, I guess we would need to change zip_set_mtime internally.

**Comment by jimhester** [comment]
Well this fixes my particular issue, but breaks the tests, will work it out.

**Comment by jimhester** [comment]
This seems sufficient to fix the installation issue and the tests are now passing. If you think more is needed @gaborcsardi you are welcome to add to this PR and merge whenever you are satisfied.

File: src/zip.c [hunk 1]
```diff
@@ -200,7 +200,11 @@ int zip_unzip(const char *czipfile, const char **cfiles, int num_files,
 
       FILE *fh = NULL;
 #ifdef _WIN32
-      fh = _wfopen(buffer, L"wb");
+      if (! cjunkpaths) {
+        fh = zip_long_wfopen(buffer, L"wb");
+      } else {
+        fh = _wfopen(buffer, L"wb");
+      }
```
Comment (gaborcsardi) [inline src/zip.c:207]: Why do we need to do this differently for junkpaths?

Comment (jimhester) [reply src/zip.c:207]: I don't know, but a test fails when we do not, perhaps the `//?/` syntax does not work with relative paths?

Comment (gaborcsardi) [reply src/zip.c:207]: Oh, of course. `\\?\` must be an absolute path, I think. So in theory we might still have an issue if the paths in the zip archive themselves are long, if `junkpaths = TRUE`. 

But I am ok with that....

Comment (jimhester) [reply src/zip.c:207]: yeah, I guess we could convert the relative path to an absolute one first, but it seems rare enough I think it is fine as-is.

**Comment by gaborcsardi** [comment]
Thanks much!

**Pull request merged**